File: zig/3-pid/Controller.zig
const std = @import("std");
const posix = std.posix;
const linux = std.os.linux;
const File = std.fs.File;
const Allocator = std.mem.Allocator;

const Registers = @import("Registers.zig");
const RingBuffer = @import("ring_buffer.zig").RingBuffer;
const pwm = @import("pwm");
const memory = @import("memory.zig");
const perf = @import("perf.zig");
const c = @import("c.zig");

const Self = @This();

const pwm_min = 0.2;
const pwm_max = 1.0;
const limit_min_deadzone = 0.001;

allocator: Allocator,
options: Options,
registers: *Registers,
i2c_file: File,
pwm_chip: *pwm.Chip,
pwm_channel: *pwm.Channel,
timer: File,
interval: struct {
    rotate_once_s: f32,
    rotate_all_s: f32,
},
state: struct {
    revolutions: RingBuffer(u32),
    is_close: bool = false,
    feedback: Feedback = .{ .delta = 0, .integration_component = 0 },
    iteration: u64 = 1,
},
perf: struct {
    read: perf.Counter,
    control: perf.Counter,
},

pub const Options = struct {
    /// Frequency of control phase, during which the following happens:
    /// * calculating the frequency for the current time window,
    /// * moving the time window forward,
    /// * updating duty cycle,
    /// * updating modbus registers.
    control_frequency: u32,
    /// Frequency is estimated for the current time window. That window is broken
    /// into [time_window_bins] bins and is moved every time the control phase
    /// takes place.
    time_window_bins: u32,
    /// Each bin in the time window gets [reads_per_bin] reads, before the next
    /// control phase fires. That means, that the read phase occurs with frequency
    /// equal to:
    ///     `control_frequency * reads_per_bin`
    /// , because every time the window moves (control phase), there must be
    /// [reads_per_bin] reads in the last bin already (read phase).
    reads_per_bin: u32,
    /// When ADC reads below this signal, the state is set to `close` to the
    /// motor magnet. If the state has changed, a new revolution is counted.
    revolution_treshold_close: u8,
    /// When ADC reads above this signal, the state is set to `far` from the
    /// motor magnet.
    revolution_treshold_far: u8,
    /// Path to I2C adapter that the ADC device is connected to.
    i2c_adapter_path: []const u8,
    /// Address on the I2C bus of the ADC device.
    i2c_address: u8,
    /// Linux PWM channel to use.
    pwm_channel: u8,
    /// Frequency of the PWM signal.
    pwm_frequency: u64,
};

const Feedback = struct { delta: f32, integration_component: f32 };

pub fn init(
    allocator: Allocator,
    registers: *Registers,
    options: Options,
) !Self {
    var revolutions = try RingBuffer(u32).init(allocator, options.time_window_bins);
    errdefer revolutions.deinit(allocator);

    const i2c_file = try std.fs.openFileAbsolute(
        options.i2c_adapter_path,
        std.fs.File.OpenFlags{ .mode = .read_write },
    );
    errdefer i2c_file.close();

    if (linux.ioctl(i2c_file.handle, c.I2C_SLAVE, options.i2c_address) < 0)
        return error.SettingI2cSlave;

    var chip = try allocator.create(pwm.Chip);
    errdefer allocator.destroy(chip);

    chip.* = try pwm.Chip.init(0);
    errdefer chip.deinit();

    var channel = try chip.channel(options.pwm_channel);
    errdefer channel.deinit();

    const timer_fd = try posix.timerfd_create(linux.CLOCK.REALTIME, .{});
    const timer = File{ .handle = timer_fd };
    errdefer timer.close();

    const read_frequency = options.control_frequency * options.reads_per_bin;
    const read_interval_us = std.time.us_per_s / read_frequency;

    const timerspec = timerspec_from_us(read_interval_us);
    try posix.timerfd_settime(timer_fd, .{}, &timerspec, null);

    try channel.setParameters(.{
        .frequency = options.pwm_frequency,
        .duty_cycle_ratio = 0.0,
    });
    try channel.enable();

    const interval_rotate_once_s: f32 =
        1.0 / @as(f32, @floatFromInt(options.control_frequency));
    const interval_rotate_all_s: f32 =
        interval_rotate_once_s * @as(f32, @floatFromInt(options.time_window_bins));

    const perf_read = try perf.Counter.init(allocator, "READ", read_frequency * 2);
    errdefer perf_read.deinit();
    const perf_control = try perf.Counter.init(allocator, "CONTROL", options.control_frequency * 2);
    errdefer perf_control.deinit();

    return .{
        .allocator = allocator,
        .options = options,
        .registers = registers,
        .i2c_file = i2c_file,
        .pwm_chip = chip,
        .pwm_channel = channel,
        .timer = timer,
        .interval = .{
            .rotate_once_s = interval_rotate_once_s,
            .rotate_all_s = interval_rotate_all_s,
        },
        .state = .{ .revolutions = revolutions },
        .perf = .{
            .read = perf_read,
            .control = perf_control,
        },
    };
}

pub fn deinit(self: *Self) void {
    self.state.revolutions.deinit(self.allocator);
    self.timer.close();
    self.pwm_channel.deinit();
    self.pwm_chip.deinit();
    self.allocator.destroy(self.pwm_chip);
    self.i2c_file.close();
}

pub const HandleResult = enum { handled, skipped };

pub fn handle(self: *Self, fd: posix.fd_t) !HandleResult {
    if (fd != self.timer.handle) return .skipped;

    var expirations: u64 = undefined;
    _ = try self.timer.readAll(std.mem.asBytes(&expirations));

    const read_start = perf.Marker.now();
    self.read_phase() catch |err| std.log.err("read_phase fail: {}", .{err});
    self.perf.read.add_sample(read_start);

    if (self.state.iteration % self.options.reads_per_bin == 0) {
        const control_start = perf.Marker.now();
        self.control_phase() catch |err| std.log.err("control_phase fail: {}", .{err});
        self.perf.control.add_sample(control_start);
    }

    const reads_per_report = self.options.reads_per_bin * self.options.control_frequency;
    if (self.state.iteration % reads_per_report == 0) {
        const report_number = self.state.iteration / reads_per_report - 1;
        std.log.info("# REPORT {}", .{report_number});
        memory.report();
        self.perf.control.report();
        self.perf.read.report();
        self.perf.control.reset();
        self.perf.read.reset();
    }

    self.state.iteration += 1;

    return .handled;
}

fn read_phase(self: *Self) !void {
    const value = try self.read_adc();

    if (!self.state.is_close and value < self.options.revolution_treshold_close) {
        // gone close
        self.state.is_close = true;
        self.state.revolutions.back().* += 1;
    } else if (self.state.is_close and value > self.options.revolution_treshold_far) {
        // gone far
        self.state.is_close = false;
    }
}

fn read_adc(self: *Self) !u8 {
    var write_value = make_read_command(0);
    var read_value: u8 = undefined;

    const addr = self.options.i2c_address;
    var msgs = [_]c.i2c_msg{
        // write command
        .{ .addr = addr, .flags = 0, .len = 1, .buf = &write_value },
        // read data
        .{ .addr = addr, .flags = c.I2C_M_RD, .len = 1, .buf = &read_value },
    };
    var data = c.i2c_rdwr_ioctl_data{ .msgs = &msgs, .nmsgs = 2 };

    const res = linux.ioctl(self.i2c_file.handle, c.I2C_RDWR, @intFromPtr(&data));
    if (res < 0)
        return posix.unexpectedErrno(posix.errno(res));

    return read_value;
}

fn make_read_command(channel: u3) u8 {
    // bit    7: single-ended inputs mode
    // bits 6-4: channel selection
    // bit    3: is internal reference enabled
    // bit    2: is converter enabled
    // bits 1-0: unused
    const default_read_command = 0b10001100;

    const channel_u8: u8 = @intCast(channel);
    return default_read_command & (channel_u8 << 4);
}

fn control_phase(self: *Self) !void {
    const frequency = self.calculate_frequency();
    self.state.revolutions.push(0);
    std.log.debug("frequency: {d:.2} Hz", .{frequency});

    const control_params = self.read_control_params();

    const control = self.calculate_control(&control_params, frequency);

    const control_signal_limited = limit(control.signal, pwm_min, pwm_max);
    std.log.debug("control signal limited: {d:.2}", .{control_signal_limited});

    try self.set_duty_cycle(control_signal_limited);

    self.write_registers(frequency, control_signal_limited);

    self.state.feedback = control.feedback;
}

fn calculate_frequency(self: *Self) f32 {
    var sum: u32 = 0;
    for (self.state.revolutions.items) |value|
        sum += value;

    return @as(f32, @floatFromInt(sum)) / self.interval.rotate_all_s;
}

pub const ControlParams = struct {
    target_frequency: f32,
    proportional_factor: f32,
    integration_time: f32,
    differentiation_time: f32,
};

fn read_control_params(self: *Self) ControlParams {
    const H = Registers.Holding;
    return .{
        .target_frequency = self.registers.get_holding(H.target_frequency),
        .proportional_factor = self.registers.get_holding(H.proportional_factor),
        .integration_time = self.registers.get_holding(H.integration_time),
        .differentiation_time = self.registers.get_holding(H.differentiation_time),
    };
}

fn calculate_control(
    self: *const Self,
    params: *const ControlParams,
    frequency: f32,
) struct { signal: f32, feedback: Feedback } {
    const interval_s = self.interval.rotate_once_s;

    const integration_factor: f32 =
        params.proportional_factor / params.integration_time * interval_s;
    const differentiation_factor: f32 =
        params.proportional_factor * params.differentiation_time / interval_s;

    const delta: f32 = params.target_frequency - frequency;
    std.log.debug("delta: {d:.2}", .{delta});

    const proportional_component: f32 = params.proportional_factor * delta;
    const integration_component: f32 =
        self.state.feedback.integration_component +
        integration_factor * self.state.feedback.delta;
    const differentiation_component: f32 =
        differentiation_factor * (delta - self.state.feedback.delta);

    const control_signal: f32 = proportional_component +
        integration_component +
        differentiation_component;

    std.log.debug("control_signal: {d:.2} = {d:.2} + {d:.2} + {d:.2}", .{
        control_signal,
        proportional_component,
        integration_component,
        differentiation_component,
    });

    return .{
        .signal = control_signal,
        .feedback = .{
            .delta = delta,
            .integration_component = integration_component,
        },
    };
}

fn set_duty_cycle(self: *Self, value: f32) !void {
    try self.pwm_channel.setParameters(.{
        .frequency = null,
        .duty_cycle_ratio = value,
    });
}

fn finite_or_zero(value: f32) f32 {
    return if (std.math.isFinite(value)) 0 else value;
}

fn limit(value: f32, min: f32, max: f32) f32 {
    if (value < limit_min_deadzone)
        return 0;

    const result = value + min;
    return if (result < min) min else if (result > max) max else result;
}

fn write_registers(self: *Self, frequency: f32, control_signal: f32) void {
    self.registers.set_input(Registers.Input.frequency, frequency);
    self.registers.set_input(Registers.Input.control_signal, control_signal);
}

fn timerspec_from_us(interval_us: u64) linux.itimerspec {
    const s = interval_us / std.time.us_per_s;
    const ns = (interval_us * std.time.ns_per_us) % std.time.ns_per_s;
    const timespec = linux.timespec{
        .tv_sec = @truncate(@as(i64, @bitCast(s))),
        .tv_nsec = @truncate(@as(i64, @bitCast(ns))),
    };
    return .{
        .it_interval = timespec,
        .it_value = timespec,
    };
}

File: rust/3-pid/src/controller.rs
use anyhow::anyhow;
use ringbuffer::{AllocRingBuffer, RingBuffer};
use rppal::i2c::I2c;
use rppal::pwm::{self, Pwm};
use std::cell::SyncUnsafeCell;
use std::sync::Arc;
use std::time::Duration;
use std::u8;
use tokio_stream::StreamExt;
use tokio_timerfd::Interval;

use crate::memory;
use crate::perf;
use crate::registers::Registers;

const PWM_MIN: f32 = 0.2;
const PWM_MAX: f32 = 1.0;

const LIMIT_MIN_DEADZONE: f32 = 0.001;

pub struct ControllerOptions {
    /// Frequency of control phase, during which the following happens:
    /// * calculating the frequency for the current time window,
    /// * moving the time window forward,
    /// * updating duty cycle,
    /// * updating modbus registers.
    pub control_frequency: u32,
    /// Frequency is estimated for the current time window. That window is broken into
    /// [time_window_bins] bins and is moved every time the control phase takes place.
    pub time_window_bins: usize,
    /// Each bin in the time window gets [reads_per_bin] reads, before the next control phase
    /// fires. That means, that the read phase occurs with frequency equal to:
    ///     `control_frequency * reads_per_bin`
    /// , because every time the window moves (control phase), there must be [reads_per_bin] reads
    /// in the last bin already (read phase).
    pub reads_per_bin: u32,
    /// When ADC reads below this signal, the state is set to `close` to the motor magnet. If the
    /// state has changed, a new revolution is counted.
    pub revolution_threshold_close: f32,
    /// When ADC reads above this signal, the state is set to `far` from the motor magnet.
    pub revolution_threshold_far: f32,
    /// Linux PWM channel to use.
    pub pwm_channel: pwm::Channel,
    /// Frequency of the PWM signal.
    pub pwm_frequency: f64,
}

pub struct Controller {
    options: ControllerOptions,
    registers: Arc<SyncUnsafeCell<Registers>>,
    pwm: Pwm,
    i2c: I2c,
    pwm_period: Duration,
    interval_rotate_once_s: f32,
    interval_rotate_all_s: f32,
    revolutions: AllocRingBuffer<u32>,
    is_close: bool,
    feedback: Feedback,
}

impl Controller {
    pub fn new(
        options: ControllerOptions,
        state: Arc<SyncUnsafeCell<Registers>>,
    ) -> anyhow::Result<Self> {
        let mut i2c = I2c::new()?;
        i2c.set_slave_address(0x48)?;

        let pwm = Pwm::new(options.pwm_channel)?;
        pwm.set_frequency(options.pwm_frequency, 0.)?;
        let pwm_period = pwm.period()?;
        pwm.enable()?;

        let mut revolutions = AllocRingBuffer::<u32>::new(options.time_window_bins);
        revolutions.fill_default();

        let interval_rotate_once_s: f32 = 1. / options.control_frequency as f32;
        let interval_rotate_all_s: f32 = interval_rotate_once_s * options.time_window_bins as f32;

        Ok(Self {
            options,
            registers: state,
            pwm,
            i2c,
            is_close: false,
            pwm_period,
            interval_rotate_once_s,
            interval_rotate_all_s,
            revolutions,
            feedback: Feedback::default(),
        })
    }

    pub async fn run(&mut self) -> anyhow::Result<!> {
        let read_frequency = self.options.control_frequency * self.options.reads_per_bin;
        let read_interval = Duration::SECOND / read_frequency;
        let mut interval = Interval::new_interval(read_interval)?;

        let perf_read_size = read_frequency as usize * 2;
        let perf_control_size = self.options.control_frequency as usize * 2;
        let mut perf_read = perf::Counter::new("READ", perf_read_size)?;
        let mut perf_control = perf::Counter::new("CONTROL", perf_control_size)?;

        let mut report_number: u64 = 0;
        while let Some(_) = interval.next().await {
            for _ in 0..self.options.control_frequency {
                for _ in 0..self.options.reads_per_bin {
                    let _read_measure = perf_read.measure();

                    if let Err(err) = self.read_phase() {
                        eprintln!("Controller::read_phase fail: {}", err);
                    }
                }

                let _control_measure = perf_control.measure();

                if let Err(err) = self.control_phase() {
                    eprintln!("Controller::control_phase fail: {}", err);
                }
            }

            println!("# REPORT {report_number}");
            memory::report();
            perf_read.report();
            perf_control.report();
            perf_read.reset();
            perf_control.reset();
            report_number += 1;
        }
        unreachable!("Interval stream is infinite");
    }

    fn read_phase(&mut self) -> anyhow::Result<()> {
        let value = self.read_adc()?;

        if value < self.options.revolution_threshold_close && !self.is_close {
            // gone close
            self.is_close = true;
            let back = self
                .revolutions
                .back_mut()
                .ok_or(anyhow!("Revolutions empty"))?;
            *back += 1;
        } else if value > self.options.revolution_threshold_far && self.is_close {
            // gone far
            self.is_close = false;
        }

        Ok(())
    }

    fn read_adc(&mut self) -> anyhow::Result<f32> {
        const WRITE_BUFFER: [u8; 1] = [adc_read_command(0)];
        let mut read_buffer = [0];

        self.i2c.write_read(&WRITE_BUFFER, &mut read_buffer)?;
        let value = read_buffer[0];
        let normalized = value as f32 / u8::MAX as f32;
        Ok(normalized)
    }

    fn control_phase(&mut self) -> anyhow::Result<()> {
        let frequency = self.calculate_frequency();
        self.revolutions.push(0);

        #[cfg(debug_assertions)]
        println!("frequency: {}", frequency);

        let registers = unsafe { self.registers.get().as_ref_unchecked() };
        let params = ControlParams::read(registers);

        let (control_signal, feedback) = self.calculate_control(&params, frequency, &self.feedback);

        let control_signal_limited = limit(control_signal, PWM_MIN, PWM_MAX);
        #[cfg(debug_assertions)]
        println!("control_signal_limited: {:.2}", control_signal_limited);
        self.write_registers(frequency, control_signal_limited);

        self.update_duty_cycle(control_signal_limited)?;
        self.feedback = feedback;

        Ok(())
    }

    fn calculate_frequency(&self) -> f32 {
        let sum: u32 = self.revolutions.iter().sum();
        sum as f32 / self.interval_rotate_all_s
    }

    fn calculate_control(
        &self,
        params: &ControlParams,
        frequency: f32,
        feedback: &Feedback,
    ) -> (f32, Feedback) {
        let delta = params.target_frequency - frequency;

        let integration_factor =
            params.proportional_factor / params.integration_time * self.interval_rotate_once_s;
        let differentiation_factor =
            params.proportional_factor * params.differentiation_time / self.interval_rotate_once_s;

        let proportional_component = params.proportional_factor * delta;
        let integration_component =
            feedback.integration_component + integration_factor * feedback.delta;
        let differentiation_component = differentiation_factor * (delta - feedback.delta);

        let control_signal =
            proportional_component + integration_component + differentiation_component;

        #[cfg(debug_assertions)]
        println!("delta: {:.2}", delta);
        #[cfg(debug_assertions)]
        println!(
            "control signal: {:.2} = {:.2} + {:.2} + {:.2}",
            control_signal,
            proportional_component,
            integration_component,
            differentiation_component
        );

        let new_feedback = Feedback {
            delta: finite_or_zero(delta),
            integration_component: finite_or_zero(integration_component),
        };
        (control_signal, new_feedback)
    }

    fn write_registers(&mut self, frequency: f32, control_signal_limited: f32) {
        let registers = unsafe { self.registers.get().as_mut_unchecked() };
        registers.write_input(.., [frequency, control_signal_limited]);
    }

    fn update_duty_cycle(&self, value: f32) -> anyhow::Result<()> {
        let pulse_width = self.pwm_period.mul_f32(value);
        self.pwm.set_pulse_width(pulse_width)?;
        Ok(())
    }
}

struct ControlParams {
    target_frequency: f32,
    proportional_factor: f32,
    integration_time: f32,
    differentiation_time: f32,
}

impl ControlParams {
    fn read(registers: &Registers) -> Self {
        let read_registers = registers.read_holding(..);

        #[rustfmt::skip]
        let &[
            target_frequency,
            proportional_factor,
            integration_time,
            differentiation_time
        ] = read_registers else { panic!("expected to read 4 registers") };

        Self {
            target_frequency,
            proportional_factor,
            integration_time,
            differentiation_time,
        }
    }
}

#[derive(Default)]
struct Feedback {
    delta: f32,
    integration_component: f32,
}

fn finite_or_zero(value: f32) -> f32 {
    if value.is_finite() {
        value
    } else {
        0.
    }
}

fn limit(signal: f32, min: f32, max: f32) -> f32 {
    if signal < LIMIT_MIN_DEADZONE {
        0.
    } else {
        (signal + min).clamp(min, max)
    }
}

const fn adc_read_command(channel: u8) -> u8 {
    assert!(channel < 8);

    // bit    7: single-ended inputs mode
    // bits 6-4: channel selection
    // bit    3: is internal reference enabled
    // bit    2: is converter enabled
    // bits 1-0: unused
    const DEFAULT_READ_COMMAND: u8 = 0b10001100;

    DEFAULT_READ_COMMAND & (channel << 4)
}
